builder/dockerfile/internals.go

```diff
@@ -0,0 +1,657 @@
+	}
+
+	config := *b.runConfig
+
+	// Create the container
+	c, warnings, err := b.docker.Create(b.runConfig, hostConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Unmount()
+	for _, warning := range warnings {
+		fmt.Fprintf(b.Stdout, " ---> [Warning] %s\n", warning)
+	}
+
+	b.tmpContainers[c.ID] = struct{}{}
+	fmt.Fprintf(b.Stdout, " ---> Running in %s\n", stringid.TruncateID(c.ID))
+
+	if config.Cmd.Len() > 0 {
+		// override the entry point that may have been picked up from the base image
+		s := config.Cmd.Slice()
+		c.Path = s[0]
+		c.Args = s[1:]
+	} else {
+		config.Cmd = stringutils.NewStrSlice()
+	}
+
+	return c, nil
+}
+
+func (b *Builder) run(c *daemon.Container) error {
+	var errCh chan error
+	if b.Verbose {
+		errCh = c.Attach(nil, b.Stdout, b.Stderr)
+	}
+
+	//start the container
+	if err := c.Start(); err != nil {
+		return err
+	}
+
+	finished := make(chan struct{})
+	defer close(finished)
+	go func() {
+		select {
+		case <-b.cancelled:
+			logrus.Debugln("Build cancelled, killing container:", c.ID)
+			c.Kill()
+		case <-finished:
+		}
+	}()
+
+	if b.Verbose {
+		// Block on reading output from container, stop on err or chan closed
+		if err := <-errCh; err != nil {
+			return err
+		}
+	}
+
+	// Wait for it to finish
+	if ret, _ := c.WaitStop(-1 * time.Second); ret != 0 {
+		// TODO: change error type, because jsonmessage.JSONError assumes HTTP
+		return &jsonmessage.JSONError{
+			Message: fmt.Sprintf("The command '%s' returned a non-zero code: %d", b.runConfig.Cmd.ToString(), ret),
+			Code:    ret,
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) clearTmp() {
+	for c := range b.tmpContainers {
+		rmConfig := &daemon.ContainerRmConfig{
+			ForceRemove:  true,
+			RemoveVolume: true,
+		}
+		if err := b.docker.Remove(c, rmConfig); err != nil {
+			fmt.Fprintf(b.Stdout, "Error removing intermediate container %s: %v\n", stringid.TruncateID(c), err)
+			return
+		}
+		delete(b.tmpContainers, c)
+		fmt.Fprintf(b.Stdout, "Removing intermediate container %s\n", stringid.TruncateID(c))
+	}
+}
+
+// readDockerfile reads a Dockerfile from the current context.
+func (b *Builder) readDockerfile() error {
+	// If no -f was specified then look for 'Dockerfile'. If we can't find
+	// that then look for 'dockerfile'.  If neither are found then default
+	// back to 'Dockerfile' and use that in the error message.
+	if b.DockerfileName == "" {
+		b.DockerfileName = api.DefaultDockerfileName
+		if _, err := b.context.Stat(b.DockerfileName); os.IsNotExist(err) {
+			lowercase := strings.ToLower(b.DockerfileName)
+			if _, err := b.context.Stat(lowercase); err == nil {
+				b.DockerfileName = lowercase
+			}
+		}
+	}
+
+	f, err := b.context.Open(b.DockerfileName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Cannot locate specified Dockerfile: %s", b.DockerfileName)
+		}
+		return err
+	}
+	if f, ok := f.(*os.File); ok {
+		// ignoring error because Open already succeeded
+		fi, _ := f.Stat()
```

Comment:
same as before, of course you opened if you have `os.File`, but `Stat` method still returns error :)